docs/public-async-nio-apis.md

```diff
@@ -0,0 +1,1162 @@
+    ///  - upgradeConfiguration: The upgrade configuration.
+    public init(httpHandlers: [NIOCore.RemovableChannelHandler], upgradeConfiguration: NIOHTTP1.NIOTypedHTTPClientUpgradeConfiguration<UpgradeResult>)
+
+    /// Called when this `ChannelHandler` is added to the `ChannelPipeline`.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ChannelHandler` belongs to.
+    public func handlerAdded(context: NIOCore.ChannelHandlerContext)
+
+    /// Called when this `ChannelHandler` is removed from the `ChannelPipeline`.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ChannelHandler` belongs to.
+    public func handlerRemoved(context: NIOCore.ChannelHandlerContext)
+
+    /// Called when the `Channel` has become active, and is able to send and receive data.
+    ///
+    /// This should call `context.fireChannelActive` to forward the operation to the next `_ChannelInboundHandler` in the `ChannelPipeline` if you want to allow the next handler to also handle the event.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ChannelHandler` belongs to.
+    public func channelActive(context: NIOCore.ChannelHandlerContext)
+
+    /// Called to request a write operation. The write operation will write the messages through the
+    /// `ChannelPipeline`. Those are then ready to be flushed to the actual `Channel` when
+    /// `Channel.flush` or `ChannelHandlerContext.flush` is called.
+    ///
+    /// This should call `context.write` to forward the operation to the next `_ChannelOutboundHandler` in the `ChannelPipeline` or
+    /// complete the `EventLoopPromise` to let the caller know that the operation completed.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ChannelHandler` belongs to.
+    ///     - data: The data to write through the `Channel`, wrapped in a `NIOAny`.
+    ///     - promise: The `EventLoopPromise` which should be notified once the operation completes, or nil if no notification should take place.
+    public func write(context: NIOCore.ChannelHandlerContext, data: NIOCore.NIOAny, promise: NIOCore.EventLoopPromise<Void>?)
+
+    /// Called when some data has been read from the remote peer.
+    ///
+    /// This should call `context.fireChannelRead` to forward the operation to the next `_ChannelInboundHandler` in the `ChannelPipeline` if you want to allow the next handler to also handle the event.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ChannelHandler` belongs to.
+    ///     - data: The data read from the remote peer, wrapped in a `NIOAny`.
+    public func channelRead(context: NIOCore.ChannelHandlerContext, data: NIOCore.NIOAny)
+
+    public func channelRead(context: NIOCore.ChannelHandlerContext, responsePart: NIOHTTP1.HTTPClientResponsePart)
```

Comment:
This presumably shouldn't be public API

Reply:
Those must be public API since they are on a public type that implements a public protocol.

Reply:
Yeah but _this_ function isn't a `ChannelHandler` method